File: src/com/l2kt/gameserver/handler/skillhandlers/Unlock.java
package com.l2kt.gameserver.handler.skillhandlers;

import com.l2kt.commons.random.Rnd;
import com.l2kt.gameserver.handler.ISkillHandler;
import com.l2kt.gameserver.model.L2Skill;
import com.l2kt.gameserver.model.WorldObject;
import com.l2kt.gameserver.model.actor.Creature;
import com.l2kt.gameserver.model.actor.ai.CtrlIntention;
import com.l2kt.gameserver.model.actor.instance.Chest;
import com.l2kt.gameserver.model.actor.instance.Door;

import com.l2kt.gameserver.network.SystemMessageId;
import com.l2kt.gameserver.network.serverpackets.SystemMessage;
import com.l2kt.gameserver.templates.skills.L2SkillType;

public class Unlock implements ISkillHandler
{
	private static final L2SkillType[] SKILL_IDS =
	{
		L2SkillType.UNLOCK,
		L2SkillType.UNLOCK_SPECIAL
	};
	
	@Override
	public void useSkill(Creature activeChar, L2Skill skill, WorldObject[] targets)
	{
		final WorldObject object = targets[0];
		
		if (object instanceof Door)
		{
			final Door door = (Door) object;
			if (!door.isUnlockable() && skill.getSkillType() != L2SkillType.UNLOCK_SPECIAL)
			{
				activeChar.sendPacket(SystemMessage.getSystemMessage(SystemMessageId.UNABLE_TO_UNLOCK_DOOR));
				return;
			}
			
			if (doorUnlock(skill) && (!door.isOpened()))
				door.openMe();
			else
				activeChar.sendPacket(SystemMessage.getSystemMessage(SystemMessageId.FAILED_TO_UNLOCK_DOOR));
		}
		else if (object instanceof Chest)
		{
			final Chest chest = (Chest) object;
			if (chest.isDead() || chest.isInteracted())
				return;
			
			chest.setInteracted();
			if (chestUnlock(skill, chest))
			{
				chest.setSpecialDrop();
				chest.doDie(null);
			}
			else
			{
				chest.addDamageHate(activeChar, 0, 999);
				chest.getAI().setIntention(CtrlIntention.ATTACK, activeChar);
			}
		}
		else
			activeChar.sendPacket(SystemMessage.getSystemMessage(SystemMessageId.INCORRECT_TARGET));
	}
	
	private static final boolean doorUnlock(L2Skill skill)
	{
		if (skill.getSkillType() == L2SkillType.UNLOCK_SPECIAL)
			return Rnd.INSTANCE.get(100) < skill.getPower();
		
		switch (skill.getLevel())
		{
			case 0:
				return false;
			case 1:
				return Rnd.INSTANCE.get(120) < 30;
			case 2:
				return Rnd.INSTANCE.get(120) < 50;
			case 3:
				return Rnd.INSTANCE.get(120) < 75;
			default:
				return Rnd.INSTANCE.get(120) < 100;
		}
	}
	
	private static final boolean chestUnlock(L2Skill skill, Creature chest)
	{
		int chance = 0;
		if (chest.getLevel() > 60)
		{
			if (skill.getLevel() < 10)
				return false;
			
			chance = (skill.getLevel() - 10) * 5 + 30;
		}
		else if (chest.getLevel() > 40)
		{
			if (skill.getLevel() < 6)
				return false;
			
			chance = (skill.getLevel() - 6) * 5 + 10;
		}
		else if (chest.getLevel() > 30)
		{
			if (skill.getLevel() < 3)
				return false;
			
			if (skill.getLevel() > 12)
				return true;
			
			chance = (skill.getLevel() - 3) * 5 + 30;
		}
		else
		{
			if (skill.getLevel() > 10)
				return true;
			
			chance = skill.getLevel() * 5 + 35;
		}
		
		chance = Math.min(chance, 50);
		return Rnd.INSTANCE.get(100) < chance;
	}
	
	@Override
	public L2SkillType[] getSkillIds()
	{
		return SKILL_IDS;
	}
}

File: src/com/l2kt/gameserver/handler/skillhandlers/Unlock.kt
package com.l2kt.gameserver.handler.skillhandlers

import com.l2kt.commons.random.Rnd
import com.l2kt.gameserver.handler.ISkillHandler
import com.l2kt.gameserver.model.L2Skill
import com.l2kt.gameserver.model.WorldObject
import com.l2kt.gameserver.model.actor.Creature
import com.l2kt.gameserver.model.actor.ai.CtrlIntention
import com.l2kt.gameserver.model.actor.instance.Chest
import com.l2kt.gameserver.model.actor.instance.Door

import com.l2kt.gameserver.network.SystemMessageId
import com.l2kt.gameserver.network.serverpackets.SystemMessage
import com.l2kt.gameserver.templates.skills.L2SkillType

class Unlock : ISkillHandler {

    override val skillIds: Array<L2SkillType>
        get() = SKILL_IDS

    override fun useSkill(activeChar: Creature, skill: L2Skill, targets: Array<WorldObject>) {
        val `object` = targets[0]

        if (`object` is Door) {
            if (!`object`.isUnlockable && skill.skillType !== L2SkillType.UNLOCK_SPECIAL) {
                activeChar.sendPacket(SystemMessage.getSystemMessage(SystemMessageId.UNABLE_TO_UNLOCK_DOOR))
                return
            }

            if (doorUnlock(skill) && !`object`.isOpened)
                `object`.openMe()
            else
                activeChar.sendPacket(SystemMessage.getSystemMessage(SystemMessageId.FAILED_TO_UNLOCK_DOOR))
        } else if (`object` is Chest) {
            if (`object`.isDead || `object`.isInteracted)
                return

            `object`.setInteracted()
            if (chestUnlock(skill, `object`)) {
                `object`.setSpecialDrop()
                `object`.doDie(null)
            } else {
                `object`.addDamageHate(activeChar, 0, 999)
                `object`.ai.setIntention(CtrlIntention.ATTACK, activeChar)
            }
        } else
            activeChar.sendPacket(SystemMessage.getSystemMessage(SystemMessageId.INCORRECT_TARGET))
    }

    companion object {
        private val SKILL_IDS = arrayOf(L2SkillType.UNLOCK, L2SkillType.UNLOCK_SPECIAL)

        private fun doorUnlock(skill: L2Skill): Boolean {
            if (skill.skillType === L2SkillType.UNLOCK_SPECIAL)
                return Rnd[100] < skill.power

            when (skill.level) {
                0 -> return false
                1 -> return Rnd[120] < 30
                2 -> return Rnd[120] < 50
                3 -> return Rnd[120] < 75
                else -> return Rnd[120] < 100
            }
        }

        private fun chestUnlock(skill: L2Skill, chest: Creature): Boolean {
            var chance = 0
            if (chest.level > 60) {
                if (skill.level < 10)
                    return false

                chance = (skill.level - 10) * 5 + 30
            } else if (chest.level > 40) {
                if (skill.level < 6)
                    return false

                chance = (skill.level - 6) * 5 + 10
            } else if (chest.level > 30) {
                if (skill.level < 3)
                    return false

                if (skill.level > 12)
                    return true

                chance = (skill.level - 3) * 5 + 30
            } else {
                if (skill.level > 10)
                    return true

                chance = skill.level * 5 + 35
            }

            chance = Math.min(chance, 50)
            return Rnd[100] < chance
        }
    }
}